validator-architecture.md

```diff
@@ -0,0 +1,164 @@
+contain.
+
+## Desired Features
+
+1. The signing service should be kept completely separate from blockchain
+   nodes. No validator private keys should ever be accessed by a component that
+has direct access to the Internet (e.g., a P2P stack).
+2. The signing service should always be protected by one or more "slash
+   protectors". This is a stateful component which stores past attestations and
+rejects any message that is slashable.
+3. The validator service should have a watchdog that provides alerts to some
+   notification service (console logs, email, etc.) if it is unable to
+obtain up-to-date blockchain information. This condition might caused by either
+node or network faults.
+4. A beacon node should be able to outsource shard syncing and block production
+   to a remote server. This avoids placing an `O(c)` bound upon the beacon
+chain node with respect to obtaining canonical shard information.
+
+## Components
+
+- **Beacon Node (Bnode)**: Provides canonical information about the beacon
+  chain.
+- **Shard Node (Snode)**: Provides canonical information about one or more shard
+  chains.
+- **Shard Node Manager (Smgr)**: Manages one or more shard nodes which
+  each sync one or more shard chains.
+- **Validator Scheduler (Vsched)**: Maintains a clock and triggers validator
+  duties like proposing a block or attesting to a shard.
+- **Validator Watchdog (Vwatch)**: Provides alerts to some external component
+  if the BeaconChain or ShardChain nodes do not provide satisfactory data.
+- **Validator Slash Protector (Vslash)**: A stateful component which stores the history of
+  a validators votes. Accepts attestations and forwards them to the Validator
+Signer if they are not slashable, otherwise discards.
+- **Validator Signer (Vsig)**: Cryptographically signs message that have passed the
+  slash protector.
+
+## Binaries
+
+Here components are placed inside binaries (listed as `exe` files, for fun).
+Three different layouts are provide to demonstrate how the component layout
+might during different phases of Serenity deployment. The `v0.0.1` phase is the
+focus of this document, future states are listed as considerations.
+
+### v0.0.1: Testnet
+
+The public testnet without sharding (e.g., Q1 2019)
+
+- **BeaconChain.exe**:
+	- BC
+	- SCMgr
+	- SC (junk data generator)
+
+- **Validator.exe**:
+	- Vsched
+	- Vwatch
+	- Vslash
+	- Vsig
+
+### v0.1.0: Testnet with Shards
+
+A public testnet with working shards.
+
+- **BeaconChain.exe**:
+	- BC
+	- SCMgr
+
+- **ShardChain.exe**:
+	- SC
+
+- **Validator.exe**:
+	- Vsched
+	- Vwatch
+	- Vslash
+	- Vsig
+
+### v1.0.0: Future
+
+Some future solution where staking hardware wallets provide portable state and
+additional slash protection.
+
+- **BeaconChain.exe**:
+	- BC
+	- SCMgr
+
+- **ShardChain.exe**:
+	- SC
+
+- **Validator.exe**:
+	- Vsched
+	- Vwatch
+	- Vslash (a)
+
+- **Staking Hardware Wallet**:
+	- Vslash (b)
+	- Vsig
+
+#### Reasoning
```

Comment:
```suggestion
## Reasoning
```
Does it make more sense for the structure to be a second level heading after `binaries`?